### PR TITLE
path expansion endpoint

### PR DESCRIPTION
**Pull request by kevinkreiser**
this adds a new endpoint to valhalla called `/expansion` which takes a normal `/route` style request but returns geojson. the geojson is a time-series dataset of which edges of the routing graph were touched by the path algorithm (only bidirectional a* implemented so far) in the order they were touched and what type of label was assigned to them.

the geojson is a small as possible (for such a verbose format). it returns only a single feature which has a geometry of type `MultiLineString`. Each `LineString` represents the beginning and end points (so only two coordinates) of each edge the algorithm touched.

the properties of the feature contain two parallel arrays (which match the number of edges/linestrings in the geometry). the first array is a list of the edge_ids and the second array is a list of the statuses. the statuses can be one of `r` (reached), `s` (settled) or `c` (connected).

this should work for multileg routes as well but there is no signifier of when one legs expansion history ends and the next one begins.

there should be little to no performance impact on the `/route` endpoint as the function that tracks this data is both `inline`d and wrapped in an `if` that says don't track this if we werent called from the `/expansion` action/endpont.

### CODE REVIEW

File: src/thor/bidirectional_astar.cc [hunk 1]
```diff
@@ -321,6 +327,10 @@ BidirectionalAStar::GetBestPath(valhalla::Location& origin,
   travel_type_ = costing_->travel_type();
   access_mode_ = costing_->access_mode();
 
+  // Set up the expansion geojson
+  if (auto expansion = expansion_.lock())
+    rapidjson::Pointer("/properties/algorithm").Set(*expansion, "bidirectional_astar");
```
Comment (kevinkreiser) [inline src/thor/bidirectional_astar.cc]: the `FeatureCollection` itself has properties which say which algorithm was used in the computation of the route

File: src/thor/route_action.cc [hunk 2]
```diff
@@ -127,6 +127,27 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+  expansion_.reset(new rapidjson::Document);
+  // every algorithm gets a default one that is empty
+  expansion_->SetObject();
+  rapidjson::Pointer("/type").Set(*expansion_, "FeatureCollection");
+  rapidjson::Pointer("/properties/algorithm").Set(*expansion_, "none");
+  rapidjson::Pointer("/features/0/type").Set(*expansion_, "Feature");
+  rapidjson::Pointer("/features/0/geometry/type").Set(*expansion_, "MultiLineString");
+  rapidjson::Pointer("/features/0/geometry/coordinates").Create(*expansion_).SetArray();
+  rapidjson::Pointer("/features/0/properties/edge_ids").Create(*expansion_).SetArray();
+  rapidjson::Pointer("/features/0/properties/statuses").Create(*expansion_).SetArray();
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: here we fill out a template of the feature collection and the one feature that holds all the linestrings. if its a multileg route we have a problem where the last leg will win on setting which algorithm wsa used. seems like in the future we should ahve a separate feature for each leg

File: src/thor/route_action.cc [hunk 3]
```diff
@@ -152,6 +173,15 @@ void thor_worker_t::route(Api& request) {
 thor::PathAlgorithm* thor_worker_t::get_path_algorithm(const std::string& routetype,
                                                        const valhalla::Location& origin,
                                                        const valhalla::Location& destination) {
+  // Tell all the algorithms where to track the expansion
+  // If the pointer has been reset it will have no effect
+  multi_modal_astar.set_expansion(expansion_);
+  timedep_forward.set_expansion(expansion_);
+  timedep_reverse.set_expansion(expansion_);
+  astar.set_expansion(expansion_);
+  bidir_astar.set_expansion(expansion_);
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: the algorithms get a weak pointer to this geojson object so they can tell that if it isnt something they should be tracking if its not backed by a real shared pointer

File: src/thor/route_action.cc [hunk 4]
```diff
@@ -127,6 +127,27 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+  rapidjson::Pointer("/features/0/geometry/type").Set(*expansion_, "MultiLineString");
+  rapidjson::Pointer("/features/0/geometry/coordinates").Create(*expansion_).SetArray();
+  rapidjson::Pointer("/features/0/properties/edge_ids").Create(*expansion_).SetArray();
+  rapidjson::Pointer("/features/0/properties/statuses").Create(*expansion_).SetArray();
+  // fill it out
+  route(request);
+  // serialize it
+  auto json = rapidjson::to_string(*expansion_, 5);
+  // tell the path algorithms to stop tracking the expansion at all
+  expansion_.reset();
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: so when we do this reset the algorithms will see that their weak_ptr doesnt point to anything and so wont track

File: valhalla/baldr/rapidjson_utils.h [hunk 5]
```diff
@@ -28,9 +28,12 @@
 
 namespace rapidjson {
 
-template <typename T> inline std::string to_string(const T& document_or_value) {
+template <typename T>
+inline std::string to_string(const T& document_or_value, int decimal_places = -1) {
   rapidjson::StringBuffer buffer;
   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  if (decimal_places >= 0)
+    writer.SetMaxDecimalPlaces(decimal_places);
```
Comment (kevinkreiser) [inline valhalla/baldr/rapidjson_utils.h]: this makes it easy for us to truncate all precision to a certain value, which is nice for geojson where we can just say we only care about 5 places for example

File: valhalla/thor/pathalgorithm.h [hunk 6]
```diff
@@ -119,9 +142,51 @@ class PathAlgorithm {
...
+   * Adds an edge to the expansion history of the current pass of the algorithm
+   * There is no schema for the status, but loosely you should pass one of:
+   * reached, settled, connected so the person using this geojson can style accordingly
+   */
+  virtual inline void TrackExpansion(baldr::GraphReader& reader,
+                                     baldr::GraphId edgeid,
+                                     const char* status,
+                                     bool full_shape = false) {
+    // only do this if someone is looking for this on the outside
+    if (auto expansion = expansion_.lock()) {
```
Comment (kevinkreiser) [inline valhalla/thor/pathalgorithm.h]: this is where we check we should actually be tracking this information or not for this run of the path algorithms getbestpath function

**Comment by kevinkreiser** [comment]
@dnesbitt61 shouldnt be any performance impact, i was very careful about that. i can do performance testing on my own but would there but a specific test suite that you would recommend to verify that?

**Comment by dnesbitt61** [comment]
I always check Lancaster to San Diego as my long test route, though any long route would work. valhalla_run_route with multi-run option works best (N successive path generation times are averaged to give the result - I usually use 100). I tried to keep a running tally of where this was, but now I have a new machine. I could try before and after with this branch but I have data without hierarchies and shortcuts.

**Comment by kevinkreiser** [comment]
I noticed a little bit more of a performance penalty locally using `std::weak_ptr` so i changed the whole thing around and decided to pass an `std::function` into the path algorithm that allows you to track what the algorithm is doing from the outside. this is cool for few reasons.

1. its more general. we are using it to do make some geojson but you could do lots of other stuff with it
2. it keeps the logic out of the path algorithm an in the action/api layer
3. its faster. in my tests there was no difference with this code and master:

```
BEFORE:
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/master$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:49:13.476930 [INFO] PathAlgorithm GetBestPath average: 120 ms
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/master$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:49:27.358655 [INFO] PathAlgorithm GetBestPath average: 119 ms
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/master$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:49:41.409925 [INFO] PathAlgorithm GetBestPath average: 119 ms
```

```
AFTER:
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/build$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:48:29.718511 [INFO] PathAlgorithm GetBestPath average: 118 ms
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/build$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:48:43.827638 [INFO] PathAlgorithm GetBestPath average: 118 ms
kkreiser@kcarbon:~/sandbox/valhalla/valhalla/build$ ./valhalla_run_route --config ../valhalla.json --multi-run 100 -j '{"locations":[{"lat":32.697958,"lon":-117.174648,"street":"Park Boulevard"},{"lat":40.042553,"lon":-76.299115}],"costing":"auto"}' | grep -F average
2019/06/14 20:48:57.987518 [INFO] PathAlgorithm GetBestPath average: 120 ms
```

File: src/thor/bidirectional_astar.cc [hunk 7]
```diff
@@ -181,6 +181,10 @@ void BidirectionalAStar::ExpandForward(GraphReader& graphreader,
                                      (pred.not_thru_pruning() || !directededge->not_thru()));
     adjacencylist_forward_->add(idx);
     *es = {EdgeSet::kTemporary, idx};
+
+    // setting this edge as reached
+    if (expansion_callback_)
```
Comment (kevinkreiser) [inline src/thor/bidirectional_astar.cc]: `std::function<>` has an `operator bool` which lets you check if its actually assigned to something callable, this check is way faster than `std::weak_ptr<>::lock`

File: src/thor/bidirectional_astar.cc [hunk 8]
```diff
@@ -754,7 +787,7 @@ std::vector<PathInfo> BidirectionalAStar::FormPath(GraphReader& graphreader) {
       cost += edgelabel.cost() - edgelabels_reverse_[predidx].cost();
     }
     cost += tc;
-    path.emplace_back(edgelabel.mode(), cost.secs, oppedge, 0, cost.cost);
+    path.emplace_back(edgelabel.mode(), cost.secs, edgelabel.opp_edgeid(), 0, cost.cost);
```
Comment (kevinkreiser) [inline src/thor/bidirectional_astar.cc]: this was a simplification. notice above that we go fetch the opposing edge id the hard way. the edge labels for bidirectional already have the opposing edge id on them so i just used that instead.  @dnesbitt61 there's not a gotcha here right?

File: src/thor/route_action.cc [hunk 9]
```diff
@@ -127,6 +128,83 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+  // default the expansion geojson so its easy to add to as we go
+  rapidjson::Document dom;
+  dom.SetObject();
+  rapidjson::Pointer("/type").Set(dom, "FeatureCollection");
+  rapidjson::Pointer("/properties/algorithm").Set(dom, "none");
+  rapidjson::Pointer("/features/0/type").Set(dom, "Feature");
+  rapidjson::Pointer("/features/0/geometry/type").Set(dom, "MultiLineString");
+  rapidjson::Pointer("/features/0/geometry/coordinates").Create(dom).SetArray();
+  rapidjson::Pointer("/features/0/properties/edge_ids").Create(dom).SetArray();
+  rapidjson::Pointer("/features/0/properties/statuses").Create(dom).SetArray();
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: so we start by prepopulating the feature

File: src/thor/route_action.cc [hunk 10]
```diff
@@ -127,6 +128,83 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+    rapidjson::Pointer("/properties/algorithm").Set(dom, algorithm);
+    rapidjson::Pointer("/features/0/properties/edge_ids")
+        .Get(dom)
+        ->GetArray()
+        .PushBack(static_cast<uint64_t>(edgeid), a);
+    rapidjson::Pointer("/features/0/properties/statuses")
+        .Get(dom)
+        ->GetArray()
+        .PushBack(rapidjson::Value{}.SetString(status, a), a);
+  };
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: we inject this lambda into the path algorithm and it calls us back with it. here we are building geojson but we could do something else with this information if we wanted.

File: src/thor/route_action.cc [hunk 11]
```diff
@@ -127,6 +128,83 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+  // tell all the algorithms how to track expansion
+  for (auto* alg : std::vector<PathAlgorithm*>{
+           &multi_modal_astar,
+           &timedep_forward,
+           &timedep_reverse,
+           &astar,
+           &bidir_astar,
+       }) {
+    alg->set_track_expansion(track_expansion);
+  }
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: this is where we tell all the algorithms to actually use our lambda to track stuff as the algorithm progresses

File: src/thor/route_action.cc [hunk 12]
```diff
@@ -127,6 +128,83 @@ void remove_edges(const GraphId& edge_id, valhalla::Location& loc, GraphReader&
...
+
+  // tell all the algorithms to stop tracking the expansion
+  for (auto* alg : std::vector<PathAlgorithm*>{
+           &multi_modal_astar,
+           &timedep_forward,
+           &timedep_reverse,
+           &astar,
+           &bidir_astar,
+       }) {
+    alg->set_track_expansion(nullptr);
```
Comment (kevinkreiser) [inline src/thor/route_action.cc]: this is where we tell all the algorithms that they no longer have a callback to call

**Pull request merged**